File: src/c/quick_find.c
#include <stdio.h>

int seq[10]; 

void init_seq(); 
void print_seq(); 
void set_union(); 

int main()
{
	init_seq(); 

	set_union(4, 3); 
	set_union(3, 8); 
	set_union(6, 5); 
	set_union(9, 4); 

	printf("%d\n", connected(8, 9)); 
	printf("%d\n", connected(5, 9)); 

	set_union(5, 0); 
	set_union(7, 2); 
	set_union(6, 1); 
	set_union(7, 3); 

	printf("%d\n", connected(2, 3)); 
	return 0; 
} 

void set_union(int p, int q)
{
	int n = seq[q];
	for(int i = 0; i < 10; i++)
	{
		if(seq[i] == n)
			seq[i] = seq[p]; 
	}
}

int connected(int p, int q)
{ 
	return seq[p] == seq[q]; 
}

void init_seq()
{
	for(int i = 0; i < 10; i++)
		seq[i] = i; 
} 

File: src/java/QuickFind.java
public class QuickFind
{
	private int[] arr = new int[10]; 

	public static void main(String[] args)
	{
		QuickFind q = new QuickFind(); 
		q.init(); 

		q.setUnion(4, 3); 
		q.setUnion(3, 8); 
		q.setUnion(6, 5); 
		q.setUnion(9, 4); 

		System.out.println(q.connected(8, 9)); 
		System.out.println(q.connected(5, 4)); 

		q.setUnion(5, 0); 
		q.setUnion(7, 2); 
		q.setUnion(6, 1); 
		q.setUnion(7, 3); 

		System.out.println(q.connected(2, 3)); 
	}

	public void init()
	{
		for(int i = 0; i < 10; i++)
			arr[i] = i; 
	}

	public void setUnion(int p, int q)
	{ 
		for(int i = 0; i < 10; i++)
		{
			if(arr[i] == arr[q])
				arr[i] = arr[p]; 
		}
	}

	public boolean connected(int p, int q)
	{
		return arr[p] == arr[q]; 
	}
}
